xarray/core/iterators.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from collections import abc
+from collections.abc import Iterator
+from typing import Callable
+
+from xarray.core.treenode import Tree
+
+"""These iterators are copied from anytree.iterators, with minor modifications."""
+
+
+class LevelOrderIter(abc.Iterator):
+    """Iterate over tree applying level-order strategy starting at `node`.
+       This is the iterator used by `DataTree` to traverse nodes.
+
+    Parameters
+    ----------
+    node : Tree
+        Node in a tree to begin iteration at.
+    filter_ : Callable, optional
+        Function called with every `node` as argument, `node` is returned if `True`.
+        Default is to iterate through all ``node`` objects in the tree.
+    stop : Callable, optional
+        Function that will cause iteration to stop if ``stop`` returns ``True``
+        for ``node``.
+    maxlevel : int, optional
+        Maximum level to descend in the node hierarchy.
+    Examples
+    --------
+    >>> from xarray.core.datatree import DataTree
+    >>> from xarray.core.iterators import LevelOrderIter
+    >>> f = DataTree(name="f")
+    >>> b = DataTree(name="b", parent=f)
+    >>> a = DataTree(name="a", parent=b)
+    >>> d = DataTree(name="d", parent=b)
+    >>> c = DataTree(name="c", parent=d)
+    >>> e = DataTree(name="e", parent=d)
+    >>> g = DataTree(name="g", parent=f)
+    >>> i = DataTree(name="i", parent=g)
+    >>> h = DataTree(name="h", parent=i)
+    >>> print(f)
+    DataTree('f', parent=None)
+    ├── DataTree('b')
+    │   ├── DataTree('a')
+    │   └── DataTree('d')
+    │       ├── DataTree('c')
+    │       └── DataTree('e')
+    └── DataTree('g')
+        └── DataTree('i')
+            └── DataTree('h')
+    >>> [node.name for node in LevelOrderIter(f)]
+    ['f', 'b', 'g', 'a', 'd', 'i', 'c', 'e', 'h']
+    >>> [node.name for node in LevelOrderIter(f, maxlevel=3)]
+    ['f', 'b', 'g', 'a', 'd', 'i']
+    >>> [
+    ...     node.name
+    ...     for node in LevelOrderIter(f, filter_=lambda n: n.name not in ("e", "g"))
+    ... ]
+    ['f', 'b', 'a', 'd', 'i', 'c', 'h']
+    >>> [node.name for node in LevelOrderIter(f, stop=lambda n: n.name == "d")]
+    ['f', 'b', 'g', 'a', 'i', 'h']
+
+    """
```

Comment:
```suggestion
    """
```

Reply:
I made the suggested whitespace changes in [this commit](https://github.com/pydata/xarray/pull/8879/commits/4ca733947cd72f2a01dd7f20b00d314600a97918).

Well I made 2.5 of them - I left the first line of the class documentation string for `LevelOrderIter` on the same line, as this seems consistent with other documentation strings in the repository. I did de-dent the second line, though, because that definitely looked a bit off.